Slice3: ExecutionHooks outcome calculation

z002Holpp opened the pull request:
Slice3: Changes for outcome calculation
Use case: In AfterHook of any hooked method we need to know the specific outcome of the hooked method. 

Main changes:
* Hook handlers have now additional parameters (method info & exception) that are needed to calculate the outcome of a hooked method
* Introduced new HookData class to encapsulate the hook handler parameters (context, method info, exception)
* Added Clone() method to TestResult in order to get a copy in the BeforeHook and diff it in the AfterHook
* Added CalculateDeltaWithPrevious() method to TestResult in order to calculate the outcome of a hooked method (difference between test result before and after the method call)

z002Holpp committed:
ExecutionHooks outcome calculation

 #revert build props

manfred-brands committed:
Small code review issues

manfred-brands committed:
Example to use ExecutionHookAttribute instead of NUnitAttribute

manfred-brands commented on src/NUnitFramework/tests/ExecutionHooks/Outcome/AfterTestHooksEvaluateTestOutcomeTests.cs:
```diff
@@ -0,0 +1,110 @@
+
+public class AfterTestHooksEvaluateTestOutcomeTests
+{
+    public class AfterTestOutcomeLogger : ExecutionHookAttribute
```
This one already nicely uses the ExecutionHookAttribute.

manfred-brands committed:
Restrict HookData to TestContext

manfred-brands reviewed (commented):
All in all this looks good.
I have pushed a few small changes.

See some comments embedded, some of them (the small ones) are addressed in my commits.

Not sure what an _Exception context_ is 
There is an Execution context, but the Exception is the result of an action.
I have dropped the _context_ from these usages.

If you can't give me a good reason, please drop `MethodInfo` from `HookData`. This then also allows dropping the `TestMethod` property on `TestMethodCommand`.

I notice all the new code directly uses the old way to add hooks instead of the `ExecutionHookAttribute`. The latter reads much cleaner. It also allows us to get async virtually for free.

Regarding your question on the `TestResult.Clone` and `CalculateDeltaWithPrevious`.
I don't mind them both being in NUnit, in case others want to use your hooks.
Although maybe not public facing.
`TestResult` even though `public` is in the `Internal` namespace.

The public facing ones are `TestContext` and `ResultAdapter` (and `TestAdapter`).
I added a commit to change HookData to use `TestContext` instead.
See if that still gives you what you need without exposing the 'Internal' namespace.

It doesn't seem any of your code needs more than `ResultAdapter`, so you could move `CalculateDeltaWithPrevious` on the adapter instead.

Finally please extract commonality from the new Outcome tests as I got a serious de-ja-vu complex here.

TravellerD14 committed:
Used ExecutionHookAttribute for attributes without test action

TravellerD14 commented on src/NUnitFramework/framework/Internal/ExecutionHooks/HookData.cs:
```diff
@@ -0,0 +1,41 @@
+        /// </summary>
+        public TestExecutionContext Context { get; }
+
+        /// <summary>
+        /// Gets the method information of the hooked method.
+        /// </summary>
+        public IMethodInfo HookedMethod { get; }
```
Please see the explanation in the [comment](https://github.com/Siemens-Healthineers/nunit/pull/27#issuecomment-3174496373)

z002Holpp committed:
fixed issues from merge conflicts

manfred-brands committed:
Ensure ExecutionHook.AddXXXHandler is internal

manfred-brands commented:
> However, one counter argument would be that `CalculateDeltaWithPrevious()` contains business logic and therefore should stay in `TestResult`, even though its not internally used and the adapter class should still only forward the call.
> 
> Would you agree with that?

As the ExecutionHooks becomes part of nunit, I'm ok with `CalculateDeltaWithPrevious` to stay on `TestResult`.
But then see my [comment](https://github.com/Siemens-Healthineers/nunit/pull/27#discussion_r2264533006) to make it work on all derived classes and probably on more fields.
I can imagine that users might want to know the delta time to report test duration as well.

manfred-brands commented:
> We could write a small and lightweight adapter class `MethodInfoAdapter` that only exposes the needed information

I like that idea. It stays on track with the other adapters hiding internals.
It allows adding more properties if needed without having to change `HookData`

Howevever `IMethodInfo` only gives the method signature, not the actual parameters it is called with.
E.g. if you have `TestCase` on method, would you not want to distinguish between 2 different calls by supplied test arguments?

There is already a `TestAdapter` which wraps several properties, including MethodName and Arguments.
Maybe you extract what you need in a `MethodCallAdapter` so it works with `SetUp`/`TearDown` and `ITestAction`.
Then `TestAdapter` can use that.

Unfortunately some PR exposed the `IMethodInfo` there and we missed that. I need to check with the PR writer why he needed it.
That property should be removed in the next major version.

z002Holpp committed:
Keeping core logic in TestResult but specific logic in derived classes.

z002Holpp committed:
Refactored CalculateDeltaWithPrevious

z002Holpp committed:
Added adapter for IMethodInfo

Integrated MethodInfoAdapter to TestContext and TestAdapter

Moved the Adapter into the TestContext

Simplified name of a type.

Removed not needed comment

Used static using to make the code more readable

Resolved small warnings regarding line endings coming from merge

Commit:
Reworked Hook Outcome Tests

 #revert build props

Commit:
small rename

TSAVogt commented on src/NUnitFramework/tests/ExecutionHooks/Outcome/AfterSetUpHooksEvaluateTestOutcomeTests.cs:
```diff
@@ -0,0 +1,158 @@
+        [SetUp]
+        public void SetUp() => ExecuteFailingReason();
+
+        private void ExecuteFailingReason()
```
Thanks a lot for the hint. When implementing the outcome tests I was focusing on the nunit test result when the different methods fail. And therefore, some outcomes differ. That was the reason why I did not put it two a common place. So, the tests had two test goals: understanding the overall nunit test outcome and the diff calculation.
I have removed the first test goal here because this is not the right place and then I was able to have much more common code in one place.

BTW: I am using a null-suppression operator in the test action outcome tests:
`ExecuteFailingReason((HookOutcomeTestsBase.FailingReason)TestContext.CurrentContext.Test.Parent!.Properties.Get(FailingReasonPropertyKey)!);`
Hope that is fine as this will bring the test to fail and code is a little bit simpler...

z002Holpp commented:
Hi Manfred,

Regarding `CalculateDeltaWithPrevious()`, 
I have done the implementation according to your proposal. For `TestSuiteResult`, I was not sure how the child elements should be handled in the delta result. We are actually not using them, so i basically clear the list of children for the delta result object.
I hope that is fine for you, if not we can still adapt it.

-------

Regarding `MethodInfo` in `HookData`, 
I introduced an adapter class in the `TestContext` ([here](https://github.com/Siemens-Healthineers/nunit/pull/27/commits/b9baf62d2c29895ca1912673f26ef09c60f74f60)) and re-used it in `HookData` such that only needed information is exposed and a user cannot (accidentally) call `Invoke()` anymore. However, I didn't find a good solution to include the parameter values also in the adapter. This is something we still have to take a closer look. I believe the workaround for us would be to take it directly from the test context.

manfred-brands committed:
Introduce a delta constructor for TestResult differences

Thsi to prevent first copying everythying and then deleting parts of it.

manfred-brands committed:
Expose less properties in TestContext.TestAdapter

Add unit tests for TestContext.MethodInfoAdapter

manfred-brands reviewed (commented):
A few comments.
I have added 2 commits:
1. Changes to determining Delta TestResult
2. Changes to TestAdapter/MethodInfoAdapter.

manfred-brands commented on src/NUnitFramework/framework/Internal/Results/TestCaseResult.cs:
```diff
@@ -67,6 +75,29 @@ public TestCaseResult(TestMethod test) : base(test)
+            {
+                StartTime = StartTime,
+                EndTime = EndTime,
+                Duration = Duration
```
Fixed in my _Introduce a delta constructor for TestResult differences_ commit.

TravellerD14 replied:
This is much better than the copy-then-delete pattern

manfred-brands commented on src/NUnitFramework/framework/Internal/Results/TestSuiteResult.cs:
```diff
@@ -230,5 +254,33 @@ private void UpdateResultState(ResultState childResultState)
+            // Therefore, we delete all the children from the delta result.
+            while (deltaResult._children.TryDequeue(out _))
+            {
+            }
```
Fixed in my _Introduce a delta constructor for TestResult differences_ commit.

z002Holpp committed:
Adapted outcome tests to change in defaut result state for test suites from NUnit (issue #5026)

manfred-brands commented on src/NUnitFramework/tests/ExecutionHooks/Outcome/AfterSetUpHooksEvaluateTestOutcomeTests.cs:
```diff
@@ -0,0 +1,158 @@
+        }
+    }
+
+    public enum FailingReason
```
That commit added a 6th copy of `FailingReason`.
And a 6th copy of "OutcomeMatched" etc.
The aim was to reduce it, not add to it.

manfred-brands reviewed (changes requested):
I unresolved my previous comment about duplication of things in tests as instead of solving it, it added another copy.
The base class is good. 
Alternatively add a static class that can be called to remove the remaining duplications.

z002Holpp committed:
Corrected some small whitespace issues that came in during merging.

z002Holpp committed:
removed old tests that have been forgotten to be deleted as part of commit "Reworked Hook Outcome Tests". This removed the duplicates of enum "FailingReason".
